### PR TITLE
Chebyshev kernel linear regression

### PR DESCRIPTION
**Pull request by MSNetrom**
Tried to implement an improved version of the "Chebyshev kernel linear regression" class from our original repository. This version uses pure pytorch to generate the data.

Class descrition:
"""
    Class for generating linear combinations of Chebyshev polynomials in given interval of degrees
    Linear combinations are generated randomly by sampling from a normal distribution
"""

Have not trained a network on this, only tested that training runs.

### CODE REVIEW

File: src/function_classes/chebyshev_specials.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        self.lowest_degree = lowest_degree
+        self.highest_degree = highest_degree
+
+        super(ChebyshevKernelLinearRegression, self).__init__(*args, **kwargs)
+
+    def _init_param_dist(self) -> D.Distribution:
+        """Produce the distribution with which to sample parameters"""
+        # Combine each polynomial randomly per sample
+        # combinations: (batch_size, 1 coefficient for each poly in chebychev_coeffs, seq_length)
+        combinations_dist = D.Normal(torch.zeros((self.x_dist.batch_shape[0], 1, self.chebyshev_coeffs.shape[0])), 1)
```
Comment (nelson-lojo) [inline src/function_classes/chebyshev_specials.py:42]: Cleanliness: You can replace `self.x_dist.batch_shape[0]` with `self.batch_size` here

File: src/function_classes/chebyshev_specials.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        # x_pows: (batch_size, seq_length, repeated x_values but to the power of different degrees)
+        x_pows = x_batch.pow(torch.arange(0, self.highest_degree + 1))
+
+        # chebychev_coeffs: (different basis polys, one element for each possible degree of poly)
+        # basis_polys: (batch_size, 1 value for each poly in chebychev_coeffs, seq_length)
+        basis_polys = self.chebyshev_coeffs @ x_pows.permute(0, 2, 1)
+
+        # Only include coefficients up to random degree
+        indices = torch.arange(0, self.chebyshev_coeffs.shape[0]).unsqueeze(0).expand(x_batch.shape[0], -1)
+        rand_tresh = torch.randint(0, self.chebyshev_coeffs.shape[0], (x_batch.shape[0], 1))
```
Comment (nelson-lojo) [inline src/function_classes/chebyshev_specials.py:60]: Cleanliness: You can replace `x_batch.shape[0]` with `self.batch_size` in both spots here

File: src/function_classes/chebyshev_specials.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        mask_indices = (rand_tresh < indices).unsqueeze(1)
+        combinations[mask_indices] = 0
+
+        # Combine basis polynomials into 1
+        return (combinations @ basis_polys).squeeze(1)
+
+class ChebychevSharedRoots(FunctionClass):
```
Comment (nelson-lojo) [inline src/function_classes/chebyshev_specials.py:67]: Should this be `ChebyshevSharedRoots`? (s, not c)

File: src/function_classes/chebyshev_specials.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    """
+    This class generates chebychev polynomials with shared roots
+    Roots can be uniformly randomly perturbed
+    """
+
+    def __init__(self, degree, perturbation=0.1, *args, **kwargs):
```
Comment (nelson-lojo) [inline src/function_classes/chebyshev_specials.py:74]: For testing, can you add a default value to `degree`?

File: src/function_classes/chebyshev_specials.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        x_batch = x_batch.expand(-1, -1, roots.shape[-1])
+
+        # Multiply factors together to get polynomial values
+        poly_values = torch.prod(x_batch - roots, dim=2)
+
+        # Add some randomness to sign, and partially random scaling
+        max_per_sample = torch.max(torch.abs(poly_values), dim=1).values
+        poly_values = poly_values * self._one_minus_one[torch.randint(0, 2, (self.batch_size, 1))] / max_per_sample.unsqueeze(1)
+
+        return poly_values
```
Comment (nelson-lojo) [inline src/function_classes/chebyshev_specials.py:108]: Cleanliness: need to add trailing newline

**Review by nelson-lojo — CHANGES_REQUESTED** [review]
A number of cleanliness changes + a spelling error

**Comment by MSNetrom** [comment]
The "chebychev shared roots" and "chebychev kernel linear regression" has the same code now.
The test-function-classes is not successfull as this function-class were made for 1-dimensional input. That is (batch_size, seq_length, 1) instead of (batch_size, seq_length, any)

**Pull request closed without merging**